lib/diaspora/federation/dispatcher.rb

```diff
@@ -0,0 +1,84 @@
+module Diaspora
+  module Federation
+    class Dispatcher
+      include Diaspora::Logging
+
+      def initialize(sender, object, opts={})
+        @sender = sender
+        @object = object
+        @opts = opts
+      end
+
+      def self.build(sender, object, opts={})
+        if object_should_be_processed_as_public?(object)
+          Public.new(sender, object, opts)
+        else
+          Private.new(sender, object, opts)
+        end
+      end
+
+      def self.defer_dispatch(sender, object, opts={})
+        object.save if object.new_record?
+        Workers::DeferredDispatch.perform_async(sender.id, object.class.to_s, object.id, opts)
+      end
+
+      def dispatch
+        deliver_to_services
+        deliver_to_subscribers
+      end
+
+      private
+
+      attr_reader :sender, :object, :opts
+
+      def self.object_should_be_processed_as_public?(object)
+        object.respond_to?(:public?) && object.public?
+      end
+      private_class_method :object_should_be_processed_as_public?
```

Comment:
`def` returns the methods name as symbol now, so in theory you could write `private_class_method def self.`, not sure we want this though.